### ISSUE TITLE
Add support for remote key/CSR generation by Venafi (service generation)

### ISSUE THREAD

**Issue by mdaugherty112** [Venafi/vault-pki-backend-venafi#57]
I'm playing around with this engine and stumbled across a few items. If I generate a certificate and key with the Venafi Vault API I get a cert and key back. The key isn't wrapped or protected with a passphrase. Also, the key isn't stored in Venafi.

Questions:
1.) Is there a way to add a passphrase to this key so it's protected at all times?
2.) Can this key somehow be stored in Venafi?

**Comment by tr1ck3r** [comment]
Repeating response to @mdaugherty112 here for community visibility:

1. We designed our secrets engine to mimic the Vault native PKI secrets engine from the clients perspective and since Vault itself doesn’t support password encrypting private keys, HashiCorp is probably the right place to start with that enhancement request.  What you’re basically asking is for them to enhance their [POST /pki/issue/:name](https://www.vaultproject.io/api-docs/secret/pki#generate-certificate) API method (and the corresponding CLI method, `vault write /pki/issue/:name`) to add an optional passphrase parameter that would password encrypt the private key before it was returned.

2. Our current Vault secrets engine implementation supports local (VCert) generated and user provided CSRs.  So one way to provide the private key visibility in TPP would be to add support for service generated keys/CSRs.  Another way would be to forward a copy of Vault generate private keys to TPP sometime during or after the enrollment process similar to what our Vault Monitoring secrets engine supports for certificates issued by Vault.  The former (service generation) is probably the most likely route we would take with the Vault (backend) secrets engine, and the latter the route we would take with our Vault Monitoring secrets engine.

This issue will be used to track the enhancement for storing private keys in Venafi.

**Comment by tr1ck3r** [comment]
I stand corrected, this solution actually already supports password based encryption for the private key by specifying the password to use when you request the certificate using the `key_password` parameter like this:
```
vault write pki-venafi/issue/tls common_name=encrypt-key.venafi.example key_password=S0m3P@ssw0rd
```
That feature was added to our project earlier this year by the work we did to achieve official "verified" status for our solution from HashiCorp. Will update the title for this issue to focus on supporting remote key/CSR generation by Venafi.

**Comment by luispresuelVenafi** [comment]
This have been addressed in [release v0.10.0](https://github.com/Venafi/vault-pki-backend-venafi/releases/tag/v0.10.0)
